pkg/virtual/framework/forwardingregistry/rest.go

```diff
@@ -0,0 +1,284 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package forwardingregistry
+
+import (
+	"context"
+	"fmt"
+
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metainternal "k8s.io/apimachinery/pkg/apis/meta/internalversion"
+	metainternalversion "k8s.io/apimachinery/pkg/apis/meta/internalversion"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/watch"
+	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/registry/rest"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/structured-merge-diff/v4/fieldpath"
+)
+
+// ClientGetter provides a way to get a dynamic client based on a given context.
+// It is used to forward REST requests to a client that depends on the request context.
+type ClientGetter interface {
+	GetDynamicClient(ctx context.Context) (dynamic.Interface, error)
+}
+
+// ForwardingREST is a REST stofzage implementation that forwards the requests to a
+// client-go dynamic client chosen based on the curret request context.
+type ForwardingREST struct {
```

Comment:
this heavily departs from the usual plumbing of REST and store objects. Wondering how we can normalize it again. Clearly, we cannot use the generic store, but we should be able to implement a forwarding store and use that instead.

Reply:
Resolved via #1037.